definitions/npm/react-redux_v4.x.x/test_react-redux_v4.x.x.js

```diff
@@ -1,65 +0,0 @@
-// @flow
-import React from 'react'
-import {connect} from 'react-redux'
-
-type Props = {
-  name: string,
-  timeOfDay: string,
-  updateTime: (time: string) => void,
-  dispatch: Function
-}
-
-class Test extends React.Component<void, Props, void> {
-  render () {
-    return <span>Hello {this.props.name}. It's {this.props.timeOfDay}</span>
-  }
-}
-
-// $ExpectError
-const a1 = <Test name='john' />; // missing props
-// $ExpectError
-const a2 = <Test name='john' timeOfDay='hg' />; // missing props
-// $ExpectError
-const a3 = <Test name='john' timeOfDay='hg' updateTime={() => undefined} />; // missing props
-
-const a4 = <Test name='john' timeOfDay='hg' updateTime={() => undefined} dispatch={() => null} />;
-// $ExpectError
-const a5 = <Test name={2345} timeOfDay='hg' updateTime={() => undefined} dispatch={() => null} />; // name must be string
-
-const ConnectedTestB = connect()(Test)
-
-// Works without dispatch. Injected implicitly
-const b1 = <ConnectedTestB name='hjh' timeOfDay='morning' updateTime={() => undefined} />;
-// $ExpectError
-const b2 = <ConnectedTestB name='hjh' timeOfDay='morning' />; // missing prop: updateTime
-// $ExpectError
-const b3 = <ConnectedTestB name={234} timeOfDay='morning' updateTime={() => undefined} />; // name: string
-
-const ConnectedTestC = connect(
-  (state: Object) => ({timeOfDay: 'connectionTime'}),
-  // (dispatch: Function) => ({
-  //   updateTime: (time: string) => undefined
-  // })
-)(Test);
-
-// Works without dispatch OR timeOfDay. Injected implicitly
-const c1 = <ConnectedTestC name='hjh' updateTime={() => undefined} />;
-// $ExpectError
-const c2 = <ConnectedTestC name='hjh' />; // missing prop: updateTime
-// $ExpectError
-const c3 = <ConnectedTestC name='345' timeOfDay={23443} updateTime={() => undefined} />; // timeOfDay: string
-
-const ConnectedTestD = connect(
-  (state: Object) => ({timeOfDay: 'connectionTime'}),
-  (dispatch: Function) => ({
-    updateTime: (time: string) => undefined,
-    dispatch
-  })
-)(Test);
-
-// Works without dispatch OR timeOfDay OR updateTime. Injected implicitly
-const d1 = <ConnectedTestD name='hjh' />;
-// $ExpectError
-const d2 = <ConnectedTestD />; // missing prop: name
-// $ExpectError
-const d3 = <ConnectedTestD name='345' timeOfDay={23443} updateTime={() => undefined} />; // timeOfDay: string
```

Comment:
@nmn thanks for reviewing this PR. Could you please elaborate a bit more? Running the new test suite against the old definition file raises 12 errors, which means there are (at least) 12 new tests and the new definition is (theorically) safer. Why do you think types are "very loose"? More details would be helpful, either to improve this definition or dismiss the changes. 

The raised errors:

```
src/test_connect.js:49
 49: // $ExpectError
     ^^^^^^^^^^^^^^^ Error suppressing comment. Unused suppression

src/test_connect.js:51
 51: // $ExpectError
     ^^^^^^^^^^^^^^^ Error suppressing comment. Unused suppression

src/test_connect.js:53
 53: // $ExpectError
     ^^^^^^^^^^^^^^^ Error suppressing comment. Unused suppression

src/test_connect.js:70
 70: // $ExpectError
     ^^^^^^^^^^^^^^^ Error suppressing comment. Unused suppression

src/test_connect.js:73
 73: // $ExpectError
     ^^^^^^^^^^^^^^^ Error suppressing comment. Unused suppression

src/test_connect.js:77
 77: // $ExpectError
     ^^^^^^^^^^^^^^^ Error suppressing comment. Unused suppression

src/test_connect.js:91
 91: // $ExpectError
     ^^^^^^^^^^^^^^^ Error suppressing comment. Unused suppression

src/test_connect.js:93
 93: // $ExpectError
     ^^^^^^^^^^^^^^^ Error suppressing comment. Unused suppression

src/test_connect.js:102
102: // $ExpectError
     ^^^^^^^^^^^^^^^ Error suppressing comment. Unused suppression

src/test_connect.js:104
104: // $ExpectError
     ^^^^^^^^^^^^^^^ Error suppressing comment. Unused suppression

src/test_connect.js:133
133: const CC8 = connector8(C3);
                 ^^^^^^^^^^^^^^ function call
 37: class C3 extends React.Component<void, Props2, void> {
                                            ^^^^^^ property `b`. Property not found in
134: <CC8 a={1} />;
     ^^^^^^^^^^^^^ props of React element `CC8`

src/test_connect.js:170
170: (CC1.WrappedComponent: C1);
          ^^^^^^^^^^^^^^^^ property `WrappedComponent`. Property not found in
170: (CC1.WrappedComponent: C1);
      ^^^ statics of React$Component
```

and the tests

```js
// @flow
import React from 'react'
import { connect } from 'react-redux'
// import type { Connector } from 'react-redux'

// copy & paste from redux libedef :(
type ReduxDispatch<A: { type: $Subtype<string> }> = (action: A) => A;

type Action = { type: 'A' } | { type: 'B' };
type Dispatch = ReduxDispatch<Action>

type State = {
  c: number,
  d: string
};

type Props1 = {
  a: number,
  b: string
};

const C1 = (props: Props1) => <div>{props.a} {props.b}</div>

type Props2 = {
  a: number,
  b: string,
  dispatch: Dispatch
};

class C2 extends React.Component<void, Props2, void> {
  render () {
    // $ExpectError
    return <div onClick={this.props.dispatch()}>{this.props.a} {this.props.b}</div>
  }
}

class C3 extends React.Component<void, Props2, void> {
  render () {
    return <div onClick={this.props.dispatch({ type: 'A' })}>{this.props.a} {this.props.b}</div>
  }
}

//
// connect()
//

const CC1 = connect()(C1);
<CC1 a={1} b="s" />;
// $ExpectError
<CC1 />; // missing a, b
// $ExpectError
<CC1 a={1} />; // missing b
// $ExpectError
<CC1 a="s" b="s" />; // wrong a type

const CC3 = connect()(C3);
<CC3 a={1} b="s" />;
// $ExpectError
<CC3 />; // missing a, b
// $ExpectError
<CC3 a={1} />; // missing b
// $ExpectError
<CC3 a="s" b="s" />; // wrong a type

//
// connect(null, null, nul, options)
//

connect(null, null, null, { pure: true })(C1);
// $ExpectError
connect(null, null, null, { pure: 1 })(C1); // wrong type
connect(null, null, null, { withRef: true })(C1);
// $ExpectError
connect(null, null, null, { withRef: 1 })(C1); // wrong type
const CC4 = connect(null, null, null, {})(C1);
<CC4 a={1} b="s" />;
// $ExpectError
<CC4 />; // missing a, b

//
// connect(mapStateToProps)
//

type OwnProps1 = { b: string };
// without ownProps
const connector5 = connect(
  (state: State) => ({ a: state.c })
);
const CC5 = connector5(C1);
<CC5 b="s" />;
// $ExpectError
<CC5 />; // missing b
// $ExpectError
<CC5 b={1} />; // wrong b type

// with ownProps
const connector6 = connect(
  (state: State, ownProps: OwnProps1) => ({ a: state.c, b: ownProps.b + 's' })
);
const CC6 = connector6(C1);
<CC6 b="s" />;
// $ExpectError
<CC6 />; // missing b
// $ExpectError
<CC6 b={1} />; // wrong b type

//
// connect(null, mapDispatchToProps)
//

type OwnProps2 = {
  a: number,
  b: string
};
// without ownProps
const connector7 = connect(
  null,
  (dispatch: Dispatch) => ({ dispatch })
);
const CC7 = connector7(C3);
<CC7 a={1} b="s" />;
// $ExpectError
<CC7 a={1} />; // missing b
// $ExpectError
<CC7 a="s" b="s" />; // wrong b type

type OwnProps3 = { a: number };
// with ownProps
const connector8 = connect(
  null,
  (dispatch: Dispatch, ownProps: OwnProps3) => ({ dispatch, b: ownProps.a + 's' })
);
const CC8 = connector8(C3);
<CC8 a={1} />;
// $ExpectError
<CC8 />; // missing a
// $ExpectError
<CC8 a="s" />; // wrong b type

//
// connect(mapStateToProps, mapDispatchToProps)
//

const connector9 = connect(
  (state: State) => ({ a: state.c }),
  (dispatch: Dispatch) => ({ dispatch })
);
const CC9 = connector9(C3);
<CC9 b="s" />;
// $ExpectError
<CC9 />; // missing b
// $ExpectError
<CC9 b={1} />; // wrong b type

//
// connect(mapStateToProps, mapDispatchToProps, MergeProps)
//

const connector10 = connect(
  (state: State) => ({ a: state.c }),
  (dispatch: Dispatch) => ({ dispatch }),
  (stateProps, dispatchProps) => Object.assign({}, stateProps, dispatchProps, { b: 's' })
);
const CC10 = connector10(C3);
<CC10 />;

//
// ConnectedComponent
//
(CC1.WrappedComponent: C1);
```